### ISSUE TITLE
Allow wildcard on src.branchName

### ISSUE THREAD

**Issue by jdimanche** [mibexsoftware/sonar-bitbucket-plugin#15]
It would be good to use wildcards for the filter to get the pull requests.
If we use the gitFlow, we create heaps of feature/* and hotfix/* branches.

Currently we must provide a branch name for PR source and it's not possible to set "feature/**".
Or at least if we could provide the dest branch, it'll be more efficient.

**Comment by mrueegg** [comment]
Hi,

The property "sonar.bitbucket.branchName" is used to determine the pull request in Bitbucket where the code review should be executed on. It is therefore not possible to use wildcards here.

Why do you need wildcards? If you use Jenkins, you can just take the build variable $GIT_BRANCH, when using Bamboo ${bamboo.repository.git.branch} to get the branch being built. That should work for all of your feature branches. Or am I missing something?

**Comment by jdimanche** [comment]
Well your plugin is usefull for applying the code analysis on the pull requests' diffs, before merging the code. So we don't need to build a version. It just scan the PR diff.

So it's not correct to get the PR from a PR source branch. It should be from a PR dest branch.

For example : I have 5 features from branches features/feature-x. All of them have a PR into the develop branch. I want that your plugin scan these PR without having to build the 5 feature branches.

That's why the wildcard could be useful. If I want to scan all the PR on all the branches, i'll set the branchName to * and it'll do the job.

**Comment by mrueegg** [comment]
I don't think this is going to work. In your example every of your 5 branches is a separate SonarQube project which means that it has to be analyzed separately. When one of your branches is built, the SonarQube plug-in only sees the issues in this specific branch/project. So you will always have to run a build/analysis for each of your branches. Or am I overlooking something?

**Comment by jdimanche** [comment]
This plugin is doing a code analysis only on the pull request diff if i'm correct.
So I don't need to build the project. I just want to have this diff analyzed to help the approval of the PR.

Today it's working with jenkins. I do not checkout the source code, so sonarqube will analyse nothing and I let your plugin connect to the bitbucket account and check for open PRs.

I have changed you plugin to remove the .filter on src.branchName and it's working (unless i have the issue #8 happening on some PRs).

**Comment by jdimanche** [comment]
It'll not be possible to have a project per source branch because when you are using the gitFlow, you create a branch per feature and you do a pull request to the develop branch. After the merge, the feature branch is deleted.

In this case i'm not able to create a project per feature branch.

**Comment by mrueegg** [comment]
Yes, you are right, the plug-in only creates issues in the diff of a pull request. But it requires that you run it on each branch separately because ultimately each branch results in its own project in SonarQube. 

If you remove the filtering on the source branch name, then the plug-in will try to create the issues on the branch it is run for all source branches of all your pull requests.

When you say "sonarqube will analyse nothing" what do you mean exactly? Our plug-in can only create issues for projects that are analyzed with SonarQube.

I'm assuming that somebody that uses this plug-in analyzes each branch separately. You can configure Jenkins to do this automatically on every new branch it recognizes (see here for example https://mibexsoftware.atlassian.net/wiki/display/SONAR4STASH/Jenkins).

**Comment by jdimanche** [comment]
I'm having a full build + sonar scan only on my consolidated develop & master branch.
I launch your plugin of all the feature branches before merging them to develop.
I won't have a sonar project per feature branches because they are temporary.

It's probably that I don't use your module the correct way. I'll use my version of your modified plugin without any filter. And if it's not what i want i'll check your link.

Thank you.

**Comment by mrueegg** [comment]
All right. What would interest me though is how you trigger SonarQube to scan your branches. Which analysis mode do you use?

**Comment by jdimanche** [comment]
sonar.analysis.mode=issues

For this analysis, i'm lauching a jenkins build and (in order to not do a full build and scan only the opened PR) I checkout no source code, so the sonarqube task will analyse nothing in the jenkins workspace.
Then your plugin is trigerred and will get all the opened PR he can have (so on all the repositories that the team have access).

This allow me to do some fast PR scan analysis only and call this build every 30 minutes (i have a lot of activity on this repositories).

**Comment by jdimanche** [comment]
I get it, I think nothing is detected if the sonar analysis is not done before your plugin is triggered.
So indeed i have to build all the src branches for each PR - i'll try your link.

**Comment by MichaelRueegg** [comment]
Just wanted to know if this worked out for you now. Please let me know so I can close this issue. Thanks in advance

**Comment by Bheesma** [comment]
On the same note - couple of questions:

- I am having sonarqube xml with all the settings mentioned in this repo. I am using MSBuild runner and Bamboo. One of the expected setting is pull request. I don't seem to find a way to get the pull request id in bamboo. Have you any details?
- I have only one project set up in sonarqube which points to my develop branch. I have configured all PR from feature branch to develop branch to be analyzed and commented on PR. I don't want separate project in sonar qube for different feature branches. Is it a valid setup supported by this tool? Also the plugin seems to be appending the pull request id to project key which does not match with the one I configured in sonarqube.
